### PR TITLE
Preparatory work for named environments

### PR DESCRIPTION
**Pull request by pmderodat**

### CODE REVIEW

File: langkit/envs.py [hunk 1]
```diff
@@ -235,25 +226,46 @@ def _parse_actions(self, actions: List[EnvAction]) -> None:
         Analyze the given list of actions and extract pre/post actions, i.e.
         actions executed before and after handling children.
         """
+        def count(cls: Type[EnvAction],
+                  sequence: Optional[List[EnvAction]] = None) -> int:
+            """
+            Return the number of ``cls`` instances in ``sequence``, or in
+            ``actions`` if ``sequence`` is left to None.
+            """
+            if sequence is None:
```
Comment (Roldak) [inline langkit/envs.py:235]: I find this added logic a bit superfluous. Simply calling `count(..., actions)` below is way shorter and even makes it easier to understand what's going on.

File: langkit/envs.py [hunk 2]
```diff
@@ -235,25 +226,46 @@ def _parse_actions(self, actions: List[EnvAction]) -> None:
...
         self.initial_env = None
         if actions and isinstance(actions[0], SetInitialEnv):
             self.initial_env = cast(SetInitialEnv, actions.pop(0))
+            first_actions.append(self.initial_env)
 
         pre, post = lsplit_by(
             lambda a: not isinstance(a, HandleChildren), actions
         )
 
+        for sequence in (pre, post):
```
Comment (Roldak) [inline langkit/envs.py:255]: why not do the check on `actions` directly? (`count(SetInitialEnv, actions) == 0`)

File: langkit/templates/astnode_types_ada.mako [hunk 3]
```diff
@@ -290,9 +292,23 @@
...
 
-   % if cls.env_spec:
+   % if cls.env_spec and cls.env_spec.actions:
 
-   <% call_prop = cls.env_spec._render_field_access %>
+   <%
+      call_prop = cls.env_spec._render_field_access
+
+      # Whether the initial env for this node is dynamic
+      has_dyn_env = cls.env_spec.initial_env
```
Comment (Roldak) [inline langkit/templates/astnode_types_ada.mako:301]: `has_dyn_env = cls.env_spec.initial_env is not None` makes it clearer IMO

**Review by Roldak — CHANGES_REQUESTED** [review]

File: langkit/envs.py (hunk 1 above)
Comment (pmderodat) [reply langkit/envs.py:235]: Good idea, thanks!

File: langkit/envs.py (hunk 2 above)
Comment (pmderodat) [reply langkit/envs.py:255]: Ah yes, you’re right. `actions` is mutated, so it looks like actually all count checks could be done directly on it. I’ll simplify this part, thanks.

File: langkit/templates/astnode_types_ada.mako (hunk 3 above)
Comment (pmderodat) [reply langkit/templates/astnode_types_ada.mako:301]: Ok, I’ll update it

**Review by Roldak — APPROVED** [review]

**Pull request merged**